python/pyarrow/tests/test_parquet.py

```diff
@@ -179,6 +179,99 @@ def alltypes_sample(size=10000, seed=0, categorical=False):
 
 
 @pytest.mark.pandas
+def test_iter_batches_columns_reader(tempdir):
+    df = alltypes_sample(size=10000, categorical=True)
+
+    filename = tempdir / 'pandas_roundtrip.parquet'
+    arrow_table = pa.Table.from_pandas(df)
+    _write_table(arrow_table, filename, version="2.0",
+                 coerce_timestamps='ms', chunk_size=1000)
+
+    columns = df.columns[4:15]
+
+    file_ = pq.ParquetFile(filename)
+
+    batches = file_.iter_batches(
+        batch_size=500,
+        columns=columns
+    )
+
+    tm.assert_frame_equal(
+        next(batches).to_pandas(),
+        df.iloc[:500, :].loc[:, columns]
+    )
+
+
+@pytest.mark.pandas
+@pytest.mark.parametrize('chunk_size', [1000])
+def test_iter_batches_reader(tempdir, chunk_size):
```

Comment:
@wjones1  thank you very much for taking over my PR. Unfortunately one of the reasons I abandoned this (and just built a wheel for my own company's project), was not knowing what to do about the problem this test highlights.

The problem with dictionary arrays is far beyond my skill. I was wondering what someone with more architectural vision thinks of this issue.

Reply:
Strangely, after I merged the latest changes from master, I am no longer seeing this issue with dictionary arrays. I definitely saw it in the original fork, so I think it may have actually been fixed (though not sure where).

I've removed the dictionary array correction in the test and hopefully the CI should confirm what I am seeing.